src/buffer/buffer_map.rs

```diff
@@ -53,57 +92,93 @@ impl BufferMap {
 pub struct IncompatibleLayout {
-    /// Names of buffers that were missing from the incoming buffer map.
-    pub missing_buffers: Vec<Cow<'static, str>>,
+    /// Identities of buffers that were missing from the incoming buffer map.
+    pub missing_buffers: Vec<BufferIdentifier<'static>>,
+    /// Identities of buffers in the incoming buffer map which cannot exist in
+    /// the target layout.
+    pub forbidden_buffers: Vec<BufferIdentifier<'static>>,
     /// Buffers whose expected type did not match the received type.
     pub incompatible_buffers: Vec<BufferIncompatibility>,
 }
 
 impl IncompatibleLayout {
-    /// Check whether a named buffer is compatible with a specific concrete message type.
-    pub fn require_message_type<Message: 'static>(
+    /// Convert this into an error if it has any contents inside.
+    pub fn as_result(self) -> Result<(), Self> {
+        if !self.missing_buffers.is_empty() {
+            return Err(self);
+        }
+
+        if !self.incompatible_buffers.is_empty() {
+            return Err(self);
+        }
+
+        Ok(())
+    }
+
+    /// Same as [`Self::require_buffer_by_literal`], but can be used with
+    /// temporary borrows of a string slice. The string slice will be cloned if
+    /// an error message needs to be produced.
+    pub fn require_buffer_by_name<BufferType: 'static>(
         &mut self,
         expected_name: &str,
         buffers: &BufferMap,
-    ) -> Result<Buffer<Message>, ()> {
-        if let Some((name, buffer)) = buffers.inner.get_key_value(expected_name) {
-            if let Some(buffer) = buffer.downcast_for_message::<Message>() {
+    ) -> Result<BufferType, ()> {
+        let identifier = BufferIdentifier::Name(Cow::Borrowed(expected_name));
+        if let Some(buffer) = buffers.get(&identifier) {
+            if let Some(buffer) = buffer.downcast_buffer::<BufferType>() {
                 return Ok(buffer);
             } else {
                 self.incompatible_buffers.push(BufferIncompatibility {
-                    name: name.clone(),
-                    expected: std::any::type_name::<Buffer<Message>>(),
+                    identifier: BufferIdentifier::Name(Cow::Owned(expected_name.to_owned())),
+                    expected: std::any::type_name::<BufferType>(),
                     received: buffer.message_type_name(),
                 });
             }
         } else {
             self.missing_buffers
-                .push(Cow::Owned(expected_name.to_owned()));
+                .push(BufferIdentifier::Name(Cow::Owned(expected_name.to_owned())));
         }
 
         Err(())
     }
 
-    /// Check whether a named buffer is compatible with a specialized buffer type,
-    /// such as `JsonBuffer`.
-    pub fn require_buffer_type<BufferType: 'static>(
+    /// Check whether a named buffer is compatible with the required buffer type.
+    pub fn require_buffer_by_literal<BufferType: 'static>(
```

Comment:
Is there a reason why we couldn't use `require_buffer_by_name`, passing it a `&'static str` literal? It doesn't seem to break anything when I tried that.

Reply:
Even if the borrow that you pass in has a static lifetime, the function can't take advantage of that, so it will need to do a heap allocation if an error occurs.